web3.js/src/util/send-and-confirm-transaction.ts

```diff
@@ -33,18 +34,36 @@ export async function sendAndConfirmTransaction(
     sendOptions,
   );
 
-  const status = (
-    await connection.confirmTransaction(
-      signature,
-      options && options.commitment,
-    )
-  ).value;
+  const status = await connection.getSignatureStatus(signature);
 
-  if (status.err) {
-    throw new Error(
-      `Transaction ${signature} failed (${JSON.stringify(status)})`,
+  const checkBlockHeight = async () => {
+    const blockHeight = await connection.getBlockHeight(
+      options && options.commitment,
     );
-  }
 
+    if (blockHeight > transaction.lastValidBlockHeight!) {
+      throw new Error('Transaction has expired.');
+    } else {
+      if (status) {
+        return;
+      }
+      await sleep(500);
+      await checkBlockHeight();
```

Comment:
This recursive approach – barring the presence of tail call optimization – will keep adding to the JavaScript call stack every time you recurse. Since the recursion here is only used for _repetition_, it's better to keep the stack shallow by using a loop.

Reply:
Refactored this a bit to use `onSignature(...)` along with a `while` loop.